### DISCUSSION TITLE
How to use markdown in R

### DISCUSSION THREAD

**Discussion by DeadlySquad13** [kiyoon/jupynium.nvim#52] [Q&A] [answered]
Hi, thanks for your plugin, enjoying it a lot!
I have a question regarding using markdown with R kernel. In R there's no block comments so recommended format from python `"""%%` throws error.

Well, it's not so convenient but I tried using line comments format like in [jupyter_ascending](https://github.com/untitled-ai/jupyter_ascending.vim):
```R
# %% [markdown]
# # Heading
# Simple Text
```
But it got rendered as:
![image](https://user-images.githubusercontent.com/46250621/221362676-6b877b05-d4e8-4032-b8fd-8f01a351ae57.png)

I have found a workaround specifying `nolint: error`:
```R
"""%% # nolint: error.
Testing out markdown
%%"""
```
but it disables errors for all file.

Please help me how to do use it in R? Maybe I'm missing some option.

**Comment by kiyoon** [comment]
Thank you. Since I don't use R I can't verify this, but can you try:

```
# %% [markdown]
# nolint start
# Heading
Simple Text
# nolint end
```

Found from https://github.com/REditorSupport/languageserver/issues/282#issuecomment-761902391

**Reply by DeadlySquad13** [reply]
> Thank you. Since I don't use R I can't verify this, but can you try:
> 
> ```
> # %% [markdown]
> # nolint start
> # Heading
> Simple Text
> # nolint end
> ```
> 
> Found from [REditorSupport/languageserver#282 (comment)](https://github.com/REditorSupport/languageserver/issues/282#issuecomment-761902391)

Thanks for reply! 
Unfortunately, this variant renders like this:
![image](https://user-images.githubusercontent.com/46250621/221364718-5c94e432-7ba4-4d35-90d7-07997ec48afb.png)

But I have found similar solution by experimenting:
```R
# nolint start
# %% [markdown]
# Title
simple text

# %%
# nolint end
Data <- read.csv("data/Computers.csv", sep = ",") 
```

Now it renders perfectly and lsp works fine. But formatting is still broken. I tried running it in console and confirmed my concerns:
```R
> style_file('PerformanceAnalysis__Lab1.ju.r')
Styling  1  files:
 PerformanceAnalysis__Lab1.ju.r ⚠
────────────────────────────────────────
Status  Count   Legend
✔       0       File unchanged.
ℹ       0       File changed.
✖       1       Styling threw an error.
────────────────────────────────────────
Warning message:
When processing PerformanceAnalysis__Lab1.ju.r: <text>:4:8: unexpected symbol
3: # Title
4: simple text
```

Styler (formatter embedded into R language server) detected wrong syntax and didn't format anything. I've found method to ignore chunks of code in [styler documentation](https://styler.r-lib.org/reference/stylerignore.html):
```diff
# nolint start
+ # styler: off
# %% [markdown]
# Title
simple text

# %%
# nolint end
+ # styler: on
Data <- read.csv("data/Computers.csv", sep = ",") 
```
However, it seems it doesn't help. I suppose this directive only ignores certain formatting issues but not syntax errors.

I think I've no ideas how to fix it on my side. Is it possible to circumvent it on plugin side?

**Reply by kiyoon** [reply]
Thanks for trying it out. I think the only good solution would be to ignore leading `# ` in markdown cells on the plugin side, just like jupyter_ascending (jupytext).

**Reply by DeadlySquad13** [reply]
Yes, it seems like a best way. I would highly appreciate if you implement it. If you're short on time, please provide some helping info on how can I do it myself.

I have also found [rmd](https://rmarkdown.rstudio.com/articles_intro.html) format. Both lsp and styler in particular have support for it and I have tested that they work fine with it. Unfortunately, it's not compatible with notebook format though cells are delimited by triple backticks so I suppose it's possible to implement a custom parser for it. But I still think that implementing workaround for plain R files is better as I don't trust nvim's treesitter performance for injected languages. Even though it was improved yesterday.

**Reply by kiyoon** [reply]
@DeadlySquad13 Done in #53! Check it out :)

**Reply by DeadlySquad13** [reply]
Thanks, it was fast! I will test it on week.

By the way, I've been digging through your code trying to implement rmd support. Even though I said it might not be the best choice performance wise, it still piqued my interest. However, I had struggles with detecting first node as it's just plain markdown without special cell indicators. I added condition to add first markdown node if the first code cell found was of rmd format (```{r}). But because this class works with partial updates, it was spamming useless markdown blocks before each new code cell. With your changes from #53 I think it now can be fixed? As I see, we can differentiate between partial buffer analyse and full buffer analyse?

**Reply by kiyoon** [reply]
It could differentiate partial update and full update already, but the difference is that, previously it didn't need to know which cell it belongs to for the partial update. After the update we can now let it know which cell type it belongs to. So if the cell type is Jupytext markdown I just removed all leading comments.

Current ju.py format parsing is like, before the first separator it is ignored (header cell). It doesn't really detect the filetype and it is an universal format for all languages.

For rmd file, I think we need to detect ```{r} and if that exists we need to set the first cell type to markdown.

**Comment by DeadlySquad13** [comment, marked as answer]
Everything works fine so far. Thanks, @kiyoon for quick update of plugin!

---
Wouldn't mark first comment as answer to avoid confusion.
